### ISSUE TITLE
[Bug] Restoring APK fails for certain apps

### ISSUE THREAD

**Issue by mvglasow** [NeoApplications/Neo-Backup#644]
**Description**
When trying to restore certain apps from backup, this fails because the APK is not found.

**Steps To Reproduce**

1. Create a backup (I did one of all user apps, APK and data)
2. Restore from that backup (I did so on a different device, as I wanted to transfer my apps with data – uninstalling individual apps or factory-resetting the device should also work)
3. Examine messages, compare apps restored to those originally installed

**Expected behavior**
All apps get restored with data.

**Actual behavior**
Some apps fail to restore, I get a notification saying

> c: base.apk is missing in backup

This affects about 20% of the apps on my device. Retrying, even with a different set of apps, does not change anything – it is always the same apps that are affected by this.

**Screenshots**
Only if there's no clear logs, add screenshots to help explain your problem.

**System Information(please complete the following information):**

- Device: OnePlusOne (backup), Fairphone 4 (restore)
- Android Version: 11 (backup), 12 (restore)
- ROM: (AOSP, MIUI, OOS...)  LineageOS 18.1 (backup), LineageOS 19 (restore)
- App's Version: 8.1.2

-----------------------------

**Additional Notes**

 If I examine the backup, I see that APK names in the folder differ. For the apps that restored properly, the APK is named `base.apk`. The others have an APK in the form `packageid-n.apk`, where `packageid` is the package ID and `n` is a number, usually 1 or 2.

Looking at the properties files, it seems that this is due to the location from which the APK was backed up, for example:

    "sourceDir":"/data/app/~~a1QHRsujsaaOOD7o5K9ltg==/at.bitfire.davdroid-WzUPLg04WmKjm4f5FySjlQ==/base.apk"

vs.

    "sourceDir":"/data/app/android.androidVNC-1.apk"

Not sure why Android would choose different names for APK files – this may have varied from version to version; this particular data history has seen Android versions from somewhere in the 4.* range to 11, over the course of 10 years. Some of the apps are updated rather frequently, others have not been updated in years.

However, the bug seems to be that, when backing up, the APK path is determined, stored in the properties file, and the APK filename in the backup is the same, with the path stripped. When restoring, however, it seems the property is completely ignored and the APK file name is assumed to always be `base.apk`.

Correct behavior would be for the restore method to read the APK path from the properties and determine the filename as the backup method does.

**Comment by machiav3lli** [comment]
I guess the architecture issue.

Split-apks are a group of small components that are unique depending on the variants that apps provide. That said Bacon is armv7(32bit) while FP4 is arm64(64bit).

**Comment by mvglasow** [comment]
I don’t think architecture is the issue here. Of two APKs that I have developed myself and which do not contain any native code, one restored fine (getting stored as base.apk) while the other one did not. Two other ones, which do contain native code and thus have some platform dependencies, restored without issues.

Switching from 32 to 64 bit should not be an issue in most cases: if the package does not contain any native code (only Dalvik/ART executables and resources), it will work in principle on any architecture. Packages that do contain native code frequently have support for the common platforms in the APK and will run on each of those platforms. (One of the apps that restore successfully is almost entirely native, with just a thin Java wrapper around, and runs fine. Since that is a project I contribute to, I know for a fact that the APK contains both arm32 and arm64 code.)

While I have not examined this in detail, the apps that fail to restore seem to be apps that I installed quite a while ago and never updated since. Those that work seem to be apps I installed more recently (I believe also those that I first installed a long time ago but have updated somewhat recently). Remember, this phone has seen many upgrades, from KitKat to 11, so APK naming may be dependent on the Android version I was on when I last upgraded (or maybe first installed) the app.

As a workaround, I opened the backup folder in the local file manager and installed the APK from that folder manually, and the apps work.

In summary, I believe it to be a genuine bug that the restore operation does not parse the APK path but seems to assume it will always be named `base.apk`.

Restoring to a device with a different architecture may be an issue if the APK does not support the new architecture. To my knowledge, the package metadata reports the supported architectures. I would expect the package manager to verify that and throw an error on mismatch – that should be easy to verify (create a backup with a package for the “wring” architecture and try to restore it, then see what happens). If the package manager fails to verify that, you could build that verification into the app – but that is optional and probably not related to what I am seeing here.

**Comment by machiav3lli** [comment]
Before we go ahead and take your assumptions for true (even apps that don't have architecture-specific code, are partitioned into a "bundle" of only system-specific components, this was done like this to save space compared to APK that contains redundant versions of components, each specific to a certain "dimension". Plus package manager doesn't give much relevant information, mostly only a parsing error), I would like you to check the content of the backup folders of the backups that don't restore and provide a view of their content.

I have two theories for this:
- Either the apps are TOO old that they only have 32bit packages, which normally shouldn't be an issue as all SD/MT SoCs still have translators for these.
- Or as already pointed out, the apps are installed as bundles and include only the 32bit libs which would result failure of installation.

To the "base.apk" thingy: The restore process depends on the backup props to take the name of the package and doesn't assume a name.

**Comment by hg42** [comment]
@machiav3lli I wonder, why you think base.apk isn't assumed:

we have
```
        const val BASE_APK_FILENAME = "base.apk"
```

so in this function:

```
    open fun restorePackage(backupDir: StorageFile, backup: Backup) {
        val packageName = backup.packageName
        Timber.i("[$packageName] Restoring from $backupDir")
        val baseApkFile = backupDir.findFile(BASE_APK_FILENAME)
            ?: throw RestoreFailedException("$BASE_APK_FILENAME is missing in backup", null)
        Timber.d("[$packageName] Found $BASE_APK_FILENAME in backup archive")
        val splitApksInBackup: Array<StorageFile> = try {
            backupDir.listFiles()
                .filter { !it.isDirectory } // Forget about dictionaries immediately
                .filter { it.name?.endsWith(".apk") == true } // Only apks are relevant
                .filter { it.name != BASE_APK_FILENAME } // Base apk is a special case
                .toTypedArray()
        } catch (e: FileNotFoundException) {
            Timber.e("Restore APKs failed: %s", e.message)
            throw RestoreFailedException(String.format("Restore APKs failed: %s", e.message), e)
        }

        // Copy all apk paths into a single array
        var apksToRestore = arrayOf(baseApkFile)
        ...
```

* the early lines throw an exception, if it doesn't find `base.apk` (q.e.d.)
* the `splitApksInBackup` exclude `base.apk` explicitly
* the `apksToRestore` starts with the found `base.apk`

then those files are copied to the staging directory as `$packageName.$apkFile` and then installed from there.

**Comment by machiav3lli** [comment]
Hehe, so my memory failed me.

**Comment by hg42** [comment]
your memory seems to be ok, but it's from the same function for restoring a system package:

They use a completely different implementation:
```
class RestoreSystemAppAction(context: Context, work: AppActionWork?, shell: ShellHandler) :
    RestoreAppAction(context, work, shell) {

    @Throws(RestoreFailedException::class)
    override fun restorePackage(backupDir: StorageFile, backup: Backup) {
        val apkTargetPath = File(backup.sourceDir ?: "")
        backupDir.findFile(apkTargetPath.name)?.let { apkFile ->
```

This leads to more questions:

* can system packages use split apks?
* shouldn't they also use the session installer?

**Comment by hg42** [comment]
> "sourceDir":"/data/app/android.androidVNC-1.apk"

I really wonder how it comes that you have a package with a file named like this in the root of `/data/app`.

I think I never saw an apk in the root of `/data/app`.
Before the scheme with the double folders with those obfuscation codes, I think we had a single folder with obfuscation, and some time ago we had simple folders with the package name only.
But no folder at all? never ever :-) though I cannot prove it... anyone with a very old phone?

**Comment by hg42** [comment]
ok, I found an old phone like this,
it uses Cyanogen 10.1 which is Android 4.2.2
and all apks are like `/data/app/$packageName-$n.apk`

**Comment by mvglasow** [comment]
Just checked my old phone, and yes, this is the exact path. The APK file has a timestamp of 2012-09-20, installed a decade and a day ago. This would presumably have been installed on CyanogenMod 9 (which is Android 4.0 “Ice Cream Sandwich”) on my old Nexus S, then copied over to the OnePlus One by backing up and restoring the `/data` partition using TWRP (both devices being on CyanogenMod 11 at the time).

The latest APKs directly in `/data/app` are from August 2015 – at that time I would have been on CM11 or 12 (Android 4.4.4 “KitKat” or 5.0 “Lollipop”). Apparently, somewhere around version 5 0r 6, Android started putting APKs into their own folders.

However, APKs located directly in `/data/app` continue to work until at least Android 11, as far as the OS is concerned.

**Comment by hg42** [comment]
we support versions down to Android 8...

but I think using the apk names listed in the properties file would be the cleaner method for all cases.
I wonder if we really need two separate functions.
The remount read/write of `/system` could be done outside of the function or may be using hook functions pre/post that could be derived instead.

**Comment by mvglasow** [comment]
> we support versions down to Android 8...

These could still have been upgraded from earlier versions and would still have the old paths, at least for apps not upgraded since.

**Comment by hg42** [comment]
we could use preprocess/postprocess, which are overridden for system apps and they are empty anyways.

but there are other differences that use some variables calculated in `restorePackage()`:
```
"(mount -o remount,rw ${quote(mountPoint)} && " +
    "mkdir -p ${quote(appDir)} && (" +  // chmod might be obsolete
    "$utilBoxQ chmod 755 ${quote(appDir)} ; " +  // for some reason a permissions error is thrown if the apk path is not created first
    "$utilBoxQ touch ${quote(apkTargetPath)} ; " + // with touch, a reboot is not necessary after restoring system apps
    "$utilBoxQ mv -f ${quote(tempPath)} ${quote(apkTargetPath)} ; " +
    "$utilBoxQ chmod 644 ${quote(apkTargetPath)}" +
    ")" +
"); mount -o remount,ro $mountPoint"
```

**Comment by hg42** [comment]
> upgraded from earlier versions and would still have the old paths

Basically I agreed above, that it could work and may be it should.

That said, I'm not sure, if the package manager would install such an old apk like a new one.
I doubt, the apk is able to control, where it is installed.
May be some backup software restored the old state exactly (Titanium?).

Could you prove that on your device? What happens, if you take such an apk and install it on a new Android version? Where is it installed?

**Comment by mvglasow** [comment]
If I install AndroidVNC on the new device using the package manager, it ends up in

    /~~EPBFBSmVhqU8yQx8BbhiMw==/android.androidVNC-t5UHVB1MemwFjXZFVR_3AQ==/base.apk

– that is, a path which follows the new conventions.

In fact, the old device has a mix of:

* APKs directly in `/data/app`
* APKs in a folder named `$packageid-$n`
* APKs in a folder named `$packageid-$randomstring==`
* APKs in a folder named `~~$randomstring==`

The new device has just the last of these four path conventions for the APK. So it seems to depend on the OS version on which the package was first installed, not on the package itself.

**Comment by hg42** [comment]
thanks, this is helpful...

so, I think, we only need to install the apks as they are found and listed in the properties file, but not assume the main apk is called base.apk

As system apks are installed by copying and not via package manager, my thought about the session installer was wrong.

If they could be split apks, may still be a valid question.

**Comment by mvglasow** [comment]
For system APKs, these are presumably updated with the OS and follow the path convention of the OS.

For user APKs, the only thing to do here is to derive the APK file name from the properties file – as far as I can tell, taking the original path from there and stripping the directories will result in the filename of the APK in the backup folder. Then install it as you always have. The package should never see the path of its own APK, thus it should not be a problem if the restore puts the APK in a different path.

**Comment by hg42** [comment]
are you member of the telegram group?

**Comment by mvglasow** [comment]
nope, I switched over from oandbackup to Neo Backup just the other day, because I needed the batch backup/restore functionality.

**Comment by hg42** [comment]
I would like to create a test apk and send it to you

**Comment by hg42** [comment]
I have an email attached to github, you may write me

**Comment by mvglasow** [comment]
Here is a backup using an ICS-style path for the APK (no significant user data, thus I can share it here). If you extract the ZIP file to the backup data folder on your device and restore, you should be able to verify it works.

[org.openintents.countdown.zip](https://github.com/NeoApplications/Neo-Backup/files/9619469/org.openintents.countdown.zip)

**Comment by hg42** [comment]
ok, thanks, though I'm not sure, if I want to install this :-) probably I shouldn't
however, I am an old guy with some trust in people, may be I will

**Comment by mvglasow** [comment]
you can compare the APK to the original to be sure it’s legit and I’m not sending you malware :-)

**Comment by hg42** [comment]
it works

**Comment by hg42** [comment]
so, either you have to wait for an official release or you can use a "pumpkin" apk (test releases from me, connotated with halloween), it's installable in parallel

**Comment by mvglasow** [comment]
I’ve worked around the issue by installing the failed APKs directly, then restoring data with Neo Backup, so I am fine for now, just thought it should be fixed on the long run. Thanks for the quick fix!
